### PR TITLE
Add support for running an A/B experiment on the new cloud cache system.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/Core/Portable/Storage/DesktopPersistenceStorageServiceFactory.cs [hunk 1]
```diff
@@ -69,6 +69,16 @@ public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
...
         }
 
+        private static StorageDatabase GetDatabase(HostWorkspaceServices workspaceServices)
+        {
+            var experimentationService = workspaceServices.GetService<IExperimentationService>();
+            if (experimentationService?.IsExperimentEnabled(WellKnownExperimentNames.CloudCache) == true)
+                return StorageDatabase.CloudCache;
+
+            var optionService = workspaceServices.GetRequiredService<IOptionService>();
+            return optionService.GetOption(StorageOptions.Database);
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Storage/DesktopPersistenceStorageServiceFactory.cs:79]: ❗ `StorageOptions.Database` needs a `LocalUserProfileStorageLocation` so the experimental path can be manually set.

**Review by sharwell — CHANGES_REQUESTED** [review]
The experimental path is missing the ability to manually set the path through vsregedit.

File: src/Workspaces/Core/Portable/Storage/DesktopPersistenceStorageServiceFactory.cs (hunk 1 above)
Comment (sharwell) [reply src/Workspaces/Core/Portable/Storage/DesktopPersistenceStorageServiceFactory.cs:79]: Take a look at the definition of `StorageOptions.SQLiteInMemoryWriteCache` prior to 44ea1444834d7563ffb1e2a30911f355a3fa8a06 for an example.

https://github.com/CyrusNajmabadi/roslyn/blob/050d32d8eeabe34653f8bf35da0fa07bb06eb067/src/Workspaces/Core/Portable/Storage/StorageOptions.cs#L23-L25

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Storage/DesktopPersistenceStorageServiceFactory.cs:79]: can do.

File: src/Workspaces/Core/Portable/Storage/StorageOptions.cs [hunk 2]
```diff
@@ -17,7 +17,9 @@ internal static class StorageOptions
 
         public const string OptionName = "FeatureManager/Storage";
 
-        public static readonly Option<StorageDatabase> Database = new(OptionName, nameof(Database), defaultValue: StorageDatabase.SQLite);
+        public static readonly Option<StorageDatabase> Database = new(
+            OptionName, nameof(Database), defaultValue: StorageDatabase.SQLite,
+            storageLocations: new LocalUserProfileStorageLocation(LocalRegistryPath + nameof(Database)));
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Storage/StorageOptions.cs:22]: ❔ Do we have other cases where an enum is stored directly? I'm assuming we do and this works but maybe a quick check to make sure?

**Review by sharwell — APPROVED** [review]

**Pull request merged**